### PR TITLE
Add value type support to DDR struct commands V2

### PR DESCRIPTION
**Pull request by tajila**
Add value type support to DDR struct commands

Add support for printing nested types in DDR. Add new command 
(!flatobject) to print nested structure. Extend existing 
!j9object command to handle nested structures.

This is based on Mark Guo's changes in #5348

Signed-off-by: tajila <atobia@ca.ibm.com>

This is a continuation from https://github.com/eclipse/openj9/pull/5878#pullrequestreview-249517520

### CODE REVIEW

**Comment by tajila** [comment]
@keithc-ca changes are ready for another review. I have tested these changes with a core file produced from a `xa6480sr5-20170810_01` VM.

Also, I updated the ValueTypeHelper functions to to address `J9FlattenedClassCache` re-structuring in https://github.com/eclipse/openj9/pull/5928.

The DTFJ API documentation is here, https://www.ibm.com/support/knowledgecenter/en/SSYKE2_8.0.0/com.ibm.java.api.80.doc/com.ibm.dtfj/index.html. None of the com.ibm.j9ddr.vm29.* classes are part of the API.

**Comment by keithc-ca** [comment]
@tajila Can you please address the merge conflicts?

**Comment by tajila** [comment]
@keithc-ca I have fixed the merge conflicts

File: debugtools/DDR_VM/src/com/ibm/j9ddr/vm29/tools/ddrinteractive/commands/FlatObjectCommand.java [hunk 1]
```diff
@@ -0,0 +1,133 @@
...
+ * FlatObjectCommand
+ *
+ * Displays all the fields of a j9object. If a field is nested, all of its fields are
+ * displayed as well. Nested fields are indented by one tab space.
+ *
+ */
+public class FlatObjectCommand extends Command {
+
+	public FlatObjectCommand() {
+		addCommand("flatobject", "<addressOfContainer> [[<fieldName1>.[<fieldName2>]...]", "Display a flattened representation of a j9object given an address.");
```
Comment (keithc-ca) [inline debugtools/DDR_VM/src/com/ibm/j9ddr/vm29/tools/ddrinteractive/commands/FlatObjectCommand.java]: Please fix nesting of `[` and `]`; I think it should be the following (note `.` has moved too).
```
<addressOfContainer> [<fieldName1>[.<fieldName2>]...]
```

File: debugtools/DDR_VM/src/com/ibm/j9ddr/vm29/tools/ddrinteractive/commands/FlatObjectCommand.java [hunk 2]
```diff
@@ -0,0 +1,133 @@
...
+	}
+
+	private void printHelp(PrintStream out) {
+		out.println("Usage: ");
+		out.println("  !flatobject <addressOfContainer> [[<fieldName1>.[<fieldName2>]...]");
```
Comment (keithc-ca) [inline debugtools/DDR_VM/src/com/ibm/j9ddr/vm29/tools/ddrinteractive/commands/FlatObjectCommand.java]: Please remove trailing space in `"Usage: "` and fix help as above.

File: debugtools/DDR_VM/src/com/ibm/j9ddr/vm29/tools/ddrinteractive/commands/FlatObjectCommand.java [hunk 3]
```diff
@@ -0,0 +1,133 @@
...
+			printHelp(out);
+			return;
+		}
+		printFlatObject(command, args, context, out);
+	}
+
+	private void printFlatObject(String command, String[] args, Context context, PrintStream out) {
+		J9ClassPointer clazz = null;
+		J9ObjectPointer object = null;
+		String argElements[] = args[0].split("\\.");
```
Comment (keithc-ca) [inline debugtools/DDR_VM/src/com/ibm/j9ddr/vm29/tools/ddrinteractive/commands/FlatObjectCommand.java]: Please rewrite this as `String[] argElements = ...`.

File: debugtools/DDR_VM/src/com/ibm/j9ddr/vm29/pointer/helper/PrintObjectFieldsHelper.java [hunk 4]
```diff
@@ -0,0 +1,321 @@
...
+import com.ibm.j9ddr.vm29.types.UDATA;
+
+
+/**
+ * Helper class to print j9object fields. Also contains helper functions
+ * for formatting output.
+ */
+public class PrintObjectFieldsHelper {
+	private static ValueTypeHelper valueTypeHelper = ValueTypeHelper.getValueTypeHelper();
+	private static final String PADDING = "\t";
```
Comment (keithc-ca) [inline debugtools/DDR_VM/src/com/ibm/j9ddr/vm29/pointer/helper/PrintObjectFieldsHelper.java]: We may decide that a tab leads to too much indenting: however, fixing that can wait.

File: debugtools/DDR_VM/src/com/ibm/j9ddr/vm29/pointer/helper/PrintObjectFieldsHelper.java [hunk 5]
```diff
@@ -0,0 +1,321 @@
...
+			}
+
+			U32 flags = new U32(J9VM_FIELD_OFFSET_WALK_INCLUDE_INSTANCE | J9VM_FIELD_OFFSET_WALK_INCLUDE_HIDDEN);
+			Iterator<J9ObjectFieldOffset> iterator = J9ObjectFieldOffsetIterator.J9ObjectFieldOffsetIteratorFor(superclass.romClass(), localClazz, previousSuperclass, flags);
+
+			while (iterator.hasNext()) {
+				J9ObjectFieldOffset result = iterator.next();
+				boolean printField = true;
+				boolean isHiddenField = result.isHidden();
+				boolean isLockword = isHiddenField && ((result.getOffsetOrAddress().add(J9Object.SIZEOF).eq(superclass.lockOffset())));
```
Comment (keithc-ca) [inline debugtools/DDR_VM/src/com/ibm/j9ddr/vm29/pointer/helper/PrintObjectFieldsHelper.java]: Please remove the two redundant sets of parentheses.

File: debugtools/DDR_VM/src/com/ibm/j9ddr/vm29/pointer/helper/PrintObjectFieldsHelper.java [hunk 6]
```diff
@@ -0,0 +1,321 @@
...
+
+				if (isLockword) {
+					/* Print the lockword field if it is indeed the lockword for this instanceClass and we haven't printed it yet. */
+					printField = !lockwordPrinted && (localClazz.lockOffset().eq(superclass.lockOffset()));
```
Comment (keithc-ca) [inline debugtools/DDR_VM/src/com/ibm/j9ddr/vm29/pointer/helper/PrintObjectFieldsHelper.java]: More redundant parentheses.

File: debugtools/DDR_VM/src/com/ibm/j9ddr/vm29/pointer/helper/ValueTypeHelper.java [hunk 7]
```diff
@@ -0,0 +1,421 @@
...
+		private MethodHandle flattenedClassCacheEntry_NaS = null;
+		private MethodHandle flattenedClassCacheEntry_clazz = null;
+		private MethodHandle flattenedClassCacheEntry_cast = null;
+		private Class<?> flattenedClassCacheEntryPointer = null;
+
+		ValueTypeSupportEnabledHelper() {
+			super();
+			try {
+				Lookup lookup = MethodHandles.lookup();
+				flattenedClassCachePointer = Class.forName("com.ibm.j9ddr.vm29.pointer.generated.J9FlattenedClassCachePointer");
```
Comment (keithc-ca) [inline debugtools/DDR_VM/src/com/ibm/j9ddr/vm29/pointer/helper/ValueTypeHelper.java]: Note: this fails in builds without support for value types; that problem will disappear once we start generating pointer classes at runtime, derived from the blob in the core file.

**Comment by tajila** [comment]
@keithc-ca I've updated the PR

File: debugtools/DDR_VM/src/com/ibm/j9ddr/vm29/pointer/helper/PrintObjectFieldsHelper.java [hunk 8]
```diff
@@ -0,0 +1,321 @@
...
+			}
+
+			U32 flags = new U32(J9VM_FIELD_OFFSET_WALK_INCLUDE_INSTANCE | J9VM_FIELD_OFFSET_WALK_INCLUDE_HIDDEN);
+			Iterator<J9ObjectFieldOffset> iterator = J9ObjectFieldOffsetIterator.J9ObjectFieldOffsetIteratorFor(superclass.romClass(), localClazz, previousSuperclass, flags);
+
+			while (iterator.hasNext()) {
+				J9ObjectFieldOffset result = iterator.next();
+				boolean printField = true;
+				boolean isHiddenField = result.isHidden();
+				boolean isLockword = isHiddenField && (result.getOffsetOrAddress().add(J9Object.SIZEOF).eq(superclass.lockOffset()));
```
Comment (keithc-ca) [inline debugtools/DDR_VM/src/com/ibm/j9ddr/vm29/pointer/helper/PrintObjectFieldsHelper.java]: Please remove the redundant (outermost) parentheses from the right operand of `&&`.

**Comment by tajila** [comment]
Change has been made

**Comment by keithc-ca** [comment]
Jenkins test sanity,extended plinux jdk11

**Pull request merged**